### PR TITLE
cancel launch permission policy update #8716

### PR DESCRIPTION
**Pull request by linuxananth1976**
@ajkerrigan @kapilt I had completed cancel ami launch permission code i.e, #8716 with test case too.  If there is anything missed out let me know.

### CODE REVIEW

**Review by kapilt — APPROVED** [review]
lgtm, I think we want to clean up the wording a little.

File: c7n/resources/ami.py [hunk 1]
```diff
@@ -295,6 +295,40 @@ def process_image(self, client, image):
...
+
+    def process(self, images):
+        client = local_session(self.manager.session_factory).client('ec2')
+        for i in images:
+            self.process_image(client, i)
+
+    def process_image(self, client, image):
+        client.cancel_image_launch_permission(
+            ImageId=image['ImageId'],
+            DryRun=self.data.get('value', True))
```
Comment (ajkerrigan) [inline c7n/resources/ami.py]: To @kapilt 's point, this won't remove permissions that come from org/OU-level shares. From some local testing, it looks like you get the same error (`InvalidAMIID.NotFound`) in all of these cases:

- The AMI doesn't exist
- The AMI exists, but is not shared with you
- The AMI exists, and is shared with you through org/OU-level permissions

I don't think we want to bomb a policy run on that exception, but I don't think we want to completely ignore it either. Perhaps a warning just to show what's going on in that case is the best we can do. Something like:

```suggestion
        try:
            client.cancel_image_launch_permission(
                ImageId=image['ImageId'],
                DryRun=self.data.get('dry_run', True),
            )
        except ClientError as err:
            if err.response["Error"]["Code"] != "InvalidAMIID.NotFound":
                raise
            self.log.warning(err.response["Error"]["Message"])
```

**Review by ajkerrigan — COMMENTED** [review]
Thanks @linuxananth1976 for working on this and getting through the initial test failures :+1: . I took it for a spin locally, have some feedback inline regarding dry runs and error handling.

File: c7n/resources/ami.py (hunk 1 above)
Comment (linuxananth1976) [reply c7n/resources/ami.py]: Thanks @ajkerrigan @kapilt  for your reviews and feedbacks, also once again thanks for @ajkerrigan applied the try exception.

Comment (kapilt) [reply c7n/resources/ami.py]: good call, re invalid id, we should handle that, got down the rabbit hole of remembering that boto ec2 doesn't have separate client exception classes for some reason.

**Review by ajkerrigan — APPROVED** [review]
Thanks for the work on this @linuxananth1976 - looks good to me :rocket:

**Pull request merged**